### PR TITLE
Update element

### PR DESCRIPTION
**Pull request by tymo**

### CODE REVIEW

File: public_html/js/component/element/FormElement.js [hunk 1]
```diff
@@ -3,36 +3,61 @@ angular.module("prosapia").factory('FormElement', function () {
...
-                    tplt = tplt.replace('<insertListemer>', element.listener).replace('<dataObject>', element.dataObject);
-                } else if (element.type === this.COMBOBOX) {
-                    tplt = tplt.replace('<model>', element.model).replace('<options>', element.options).replace('<option>', element.option);
-                } else if ([this.TEXTINPUT, this.DATEINPUT, this.DATEPICKER].includes(element.type)) {
-                    tplt = tplt.replace('<name>', element.name).replace('<model>', element.model).replace('<placeholder>', element.placeholder);
-                    tplt += "<br>";
+            })[0].tag : null);
+            newElement = null;
+            if (tag) {
+                newElement = document.createElement(tag);
```
Comment (flachadriano) [inline public_html/js/component/element/FormElement.js]: As we discussed, the method to create the attributes is being very big and the logic for all components are making bigger the method.

So, in Tasy, we create a Factory for each component, defining what it is necessary to build it.

File: public_html/js/component/element/FormElement.js [hunk 2]
```diff
@@ -1,38 +1,30 @@
...
+
         elements: [
-            {type: 'TXT', content: '<input type="text" name="<name>" ng-model="data.<model>" placeholder="<placeholder>" class="textInput"/>'},
-            {type: 'TXD', content: '<input type="text" ng-model="data.<model>" name="<name>" placeholder="<placeholder>" ui-date/>'},
-            {type: 'DTP', content: '<input type="date" name="<name>" ng-model="data.<model>" placeholder="<placeholder>" class="textInput" />'},
-            {type: 'BTN', content: '<button class="addButton" name="sendButton" ng-click="addItem(listName, data)">Adicionar</button>'},
-            {type: 'SLT', content: '<select class="form-control" ng-model="data.<model>" ng-options="<options>">\
-             <option value=""><option></option>\
-             </select>'}
+            {type: 'TXT', cmp: TextInput},
```
Comment (flachadriano) [inline public_html/js/component/element/FormElement.js]: This need to be a array of objects?

Could we transform it in a unique object?

We have only two infos that changes in this array of objects.

Why not transform this in a object key-value?

File: public_html/js/component/element/FormElement.js [hunk 3]
```diff
@@ -1,38 +1,30 @@
...
+            {type: 'BTN', cmp: Button},
+            {type: 'SLT', cmp: ListBox}
         ],
-        getElement: function (element) {
-            let tplt = (this.elements.filter(function (cp) {
-                return cp.type === element.type;
+
+        getElement: function (elemInfo) {
+            let element = (this.elements.filter(function (cp) {
```
Comment (flachadriano) [inline public_html/js/component/element/FormElement.js]: Transform it based on the new structure of the object this.elements

File: public_html/js/controller/prosapiaCtrl.js [hunk 4]
```diff
@@ -3,14 +3,26 @@ angular.module("prosapia").controller("prosapiaCtrl", function ($scope, FormElem
     $scope.app = "Prosapia";
 
     $scope.dosageInputFields = [
-        {type: FormElement.TEXTINPUT, name: 'name', model: 'name', placeholder: 'Nome'},
-        {type: FormElement.BUTTONSUBMIT, listName: 'dosageList'}
+        {type: FormElement.TEXTINPUT, name: 'name', model: 'name', placeHolder: 'Nome', eType: 'text'},
```
Comment (flachadriano) [inline public_html/js/controller/prosapiaCtrl.js]: Create a Factory Builder for to create this object.

In the end, should looks like:
Input.name('name').model('name').placeHolder('Nome').eType('text').build()

Attention: add on method build the consistences to create the object. If the object does not have a mandatory info, throw exception.

File: public_html/js/controller/prosapiaCtrl.js [hunk 5]
```diff
@@ -3,14 +3,26 @@ angular.module("prosapia").controller("prosapiaCtrl", function ($scope, FormElem
...
+        {type: FormElement.BUTTONSUBMIT, class: "addButton", click: "addItem(\'dosageList\', data, $event)"}
     ];
 
     $scope.medicineInputFields = [
-        {type: FormElement.TEXTINPUT, name: 'name', model: 'name', placeholder: 'Nome'},
-        {type: FormElement.COMBOBOX, model: 'dosage', options: 'dosage.name for dosage in List.getList(\'dosageList\') track by dosage.id', option: 'Selecione a forma de dosagem'},
-        {type: FormElement.BUTTONSUBMIT, listName: 'medicineList'}
+        {type: FormElement.TEXTINPUT, name: 'name', model: 'name', placeHolder: 'Nome', eType: 'text'},
+        {type: FormElement.SELECT, model: 'dosage', options: 'dosage.name for dosage in List.getList(\'dosageList\') track by dosage.id', option: 'Selecione a forma de dosagem'},
```
Comment (flachadriano) [inline public_html/js/controller/prosapiaCtrl.js]: Transform the object options into separated parameters.

The development does need to understand the Angular structure.

What if we need to change the Framework? They all will need to change the structures?

File: public_html/js/controller/prosapiaCtrl.js [hunk 6]
```diff
@@ -3,14 +3,26 @@ angular.module("prosapia").controller("prosapiaCtrl", function ($scope, FormElem
...
+        {type: FormElement.BUTTONSUBMIT, class: "addButton", click: "addItem(\'medicineList\', data, $event)"}
```
Comment (flachadriano) [inline public_html/js/controller/prosapiaCtrl.js]: In all BUTTONSUBMIT the developer must sent the class and click action?

How could we transform it into variables? Or we even need that?

File: public_html/js/component/element/ListBox.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+angular.module("prosapia").factory('ListBox', function () {
+    function build(element) {
+        newElement = document.createElement('SELECT');
```
Comment (flachadriano) [inline public_html/js/component/element/ListBox.js]: NEXT STEP

Create a directive for the Select component.

So, this factory will be responsible for create a the new component.

In this new component we will have how to create a Select component.

This directive will have specific CSS (if necessary) and HTML

File: public_html/js/component/element/FormElement.js (hunk 2 above)
Comment (tymo) [reply public_html/js/component/element/FormElement.js]: Updated, now the type refers directly to the element factory wich is detected on a switch/case block,

File: public_html/js/component/element/FormElement.js (hunk 3 above)
Comment (tymo) [reply public_html/js/component/element/FormElement.js]: Done, now thare's a switch that creates the specified element.

**Pull request merged**